packages/typespec-vscode/src/npm-utils.ts

```diff
@@ -0,0 +1,143 @@
+import fs from "fs";
+import path from "path";
+import semver from "semver";
+import logger from "./log/logger.js";
+import { ExecOutput, loadModule, spawnExecution, spawnExecutionEvents } from "./utils.js";
+
+export enum InstallationAction {
+  Install = "Install",
+  Upgrade = "Upgrade",
+  Skip = "Skip",
+  Cancel = "Cancel",
+}
+
+export enum npmDependencyType {
+  dependencies = "dependencies",
+  peerDependencies = "peerDependencies",
+  devDependencies = "devDependencies",
+}
+
+export interface NpmPackageInfo {
+  name: string;
+  version?: string;
+  resolved?: string;
+  overridden?: string;
+  dependencies?: Record<string, NpmPackageInfo>;
+}
+
+export class NpmUtil {
+  private cwd: string;
+
+  constructor(cwd: string) {
+    this.cwd = cwd;
+  }
+
+  public async npmInstallPackages(
+    packages: string[] = [],
+    options: any = {},
+    on?: spawnExecutionEvents,
+  ): Promise<ExecOutput> {
+    return spawnExecution("npm", ["install", ...packages], this.cwd, on);
+  }
+
+  public async ensureNpmPackageInstall(
+    packageName: string,
+    version?: string,
+  ): Promise<{ action: InstallationAction; version?: string }> {
+    const { installed: isPackageInstalled, version: installedVersion } =
+      await this.isPackageInstalled(packageName);
+    if (isPackageInstalled) {
+      if (version && installedVersion !== version) {
+        return {
+          action: InstallationAction.Upgrade,
```

Comment:
what if downgrade?

Reply:
if downgrade, we will skip install and log

Reply:
but you only compared they are not equal, how would you determine it must be upgrade here?

Reply:
I compared those two version, greater or less. I updated the code.